Source/CesiumRuntime/Private/CesiumTextureUtility.cpp

```diff
@@ -1,18 +1,351 @@
+
+      FCesiumTextureData bulkData(*_pCesiumImage);
+
+      FRHIResourceCreateInfo createInfo{};
+      createInfo.BulkData = &bulkData;
+      createInfo.ExtData = _platformExtData;
+
+      ETextureCreateFlags textureFlags = TexCreate_ShaderResource;
+
+      if (this->bSRGB) {
+        textureFlags |= TexCreate_SRGB;
+      }
+
+      FTexture2DRHIRef rhiTexture;
+
+      // Copies over mip0, allocates the rest of the mips if needed.
+
+      // RHICreateTexture2D can actually copy over all the mips in one shot,
+      // but it expects a particular memory layout. Might be worth configuring
+      // Cesium Native's mip-map generation to obey a standard memory layout.
+      rhiTexture = RHICreateTexture2D(
+          static_cast<uint32>(_pCesiumImage->width),
+          static_cast<uint32>(_pCesiumImage->height),
+          _format,
+          _generateMipMaps
+              ? static_cast<uint32>(_pCesiumImage->mipPositions.size())
+              : 1,
+          1,
+          textureFlags,
+          createInfo);
+
+      // Copies over rest of the mips
+      if (_generateMipMaps) {
+        for (uint32 i = 1;
+             i < static_cast<uint32>(_pCesiumImage->mipPositions.size());
+             ++i) {
+          uint32 DestPitch;
+          void* pDestination =
+              RHILockTexture2D(rhiTexture, i, RLM_WriteOnly, DestPitch, false);
+          CopyMip(pDestination, DestPitch, _format, *_pCesiumImage, i);
+          RHIUnlockTexture2D(rhiTexture, i, false);
+        }
+      }
+
+      this->TextureRHI = rhiTexture;
+      rhiTexture.SafeRelease();
+    }
+
+    RHIUpdateTextureReference(TextureReferenceRHI, this->TextureRHI);
+  }
+
+  virtual void ReleaseRHI() override {
+    RHIUpdateTextureReference(TextureReferenceRHI, nullptr);
+    TextureReferenceRHI.SafeRelease();
+    TextureRHI.SafeRelease();
+
+    FTextureResource::ReleaseRHI();
+  }
+
+private:
+  UTexture* _pTexture;
+  // TODO: remove this, temporary workaround
+  CesiumGltf::ImageCesium _image;
+  const CesiumGltf::ImageCesium* _pCesiumImage;
+  EPixelFormat _format;
+  bool _generateMipMaps;
+  uint32 _platformExtData;
+};
+
+/**
+ * @brief Create an RHI texture on this thread. This requires
+ * GRHISupportsAsyncTextureCreation to be true.
+ *
+ * @param image The CPU image to create on the GPU.
+ * @param format The pixel format of the image.
+ * @param generateMipMaps Whether the RHI texture should have a mipmap.
+ * @return The RHI texture reference.
+ */
+FTexture2DRHIRef CreateRHITexture2D_Async(
+    const CesiumGltf::ImageCesium& image,
+    EPixelFormat format,
+    bool generateMipMaps) {
+  check(GRHISupportsAsyncTextureCreation);
+
+  ETextureCreateFlags textureFlags = TexCreate_ShaderResource | TexCreate_SRGB;
+
+  if (generateMipMaps) {
+    uint32 mipCount = static_cast<uint32>(image.mipPositions.size());
+
+    // TODO
+    // find way to reuse this heap allocation, static thread local??
+    std::vector<void*> mipsData(mipCount);
```

Comment:
mipCount will realistically never be larger than... what?... 16 maybe? So I'd just use a stack allocated array:

```
void* mipsData[16];
```

Reply:
Used your suggestion here, good idea!